src/interfaces/controllers/deregister-trailer/deregister-trailer.ts

```diff
@@ -0,0 +1,54 @@
+import { Request, Response, NextFunction } from 'express';
+import * as domain from '../../../domain';
+import { DataAccess } from '../../../utils/data-access';
+import { IRequestHandler } from '../i-request-handler';
+import { deregisterTrailerValidator } from './deregister-trailer-validator';
+
+export class DeregisterTrailer implements IRequestHandler {
+  private validate(payload?: domain.DeregisterTrailer): string {
+    const errors = deregisterTrailerValidator.validate(payload, { abortEarly: false }).error;
+    return errors?.message;
+  }
+
+  public async call(req: Request, res: Response, next: NextFunction): Promise<void> {
+    const deregisterTrailer = req.body as domain.DeregisterTrailer;
+    const { trn } = req.params;
+
+    const errors = this.validate(deregisterTrailer);
+    if (errors) {
+      next(new domain.HTTPError(400, errors));
+      return;
+    }
+
+    try {
+      const { reasonForDeregistration, deregisterDate } = deregisterTrailer;
+      let existingTrailerRegistration = await this.checkForExistingTrailerRegistration(trn);
+      if (existingTrailerRegistration) {
+        existingTrailerRegistration = {
+          ...existingTrailerRegistration,
+          reasonForDeregistration,
+          deregisterDate,
+        };
+        await DataAccess.getInstance().upsertTrailerRegistration(existingTrailerRegistration);
+      }
+    } catch (error) {
+      next(new domain.HTTPError(500, error));
+      return;
+    }
+    res.status(204).send();
+  }
+
+  private async checkForExistingTrailerRegistration(trn: string): Promise<domain.TrailerRegistration | null> {
+    const trailerRegistration = await DataAccess.getInstance().getByTrn(trn);
+    if (!trailerRegistration) {
+      return null;
```

Comment:
Open question on the requirement here as I believe we still return 204 which is success when the trn is not found. Don't we want to tell the user that the trn is not found still??
Basically the AC4 and AC5 return the same status code on the same endpoint for 2 possible outcome use cases, so how would the user know? Or do they have know?